src/crates.rs

```diff
@@ -0,0 +1,277 @@
+use colored::Colorize;
+use hmac_sha256::Hash;
+use std::{
+    fmt::Write,
+    fs::DirEntry,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+/// Represents a handle to a .crate file that has been downloaded to disk
+/// to a temporary folder. The temporary file and folder will be dropped
+/// once this goes out of scope.
+/// This retains a reference to the Config it was downloaded from, so it
+/// is tied to it's lifetime.
+pub struct CrateFile<'a> {
+    crate_file_prefix: String,
+    crate_file_extension: &'static str,
+    tempdir_handle: tempfile::TempDir,
+    config: &'a Config,
+}
+
+/// Represents a handle to a .crate file that has been downloaded to disk
+/// to a temporary folder, extracted, and built in release mode.
+/// The temporary file(s) and folder will be dropped once this goes out of scope.
+/// This retains a reference to the Config it was built from, so it
+/// is tied to it's lifetime.
+pub struct BuiltCrate<'a> {
+    crate_file_prefix: String,
+    tempdir_handle: tempfile::TempDir,
+    musl: bool,
+    config: &'a Config,
+}
+
+impl<'a> CrateFile<'a> {
+    /// Download the `.crate` file from `crates.io` to a temporary directory,
+    /// and if succesful, return a handle to it.
+    pub fn download_new(config: &Config) -> Result<CrateFile, Error> {
+        let pkgname = &config.package.name;
+        let pkgver = &config.package.version;
+        // This is downloaded to a temporary directory, instead of the current
+        // directory, to avoid the issue where cargo thinks its in a workspace.
+        // https://github.com/rust-lang/cargo/issues/5418
+        let tempdir_handle = tempfile::tempdir()?;
+        let crate_file_prefix = format!("{pkgname}-{pkgver}");
+        let crate_file_extension = "tar.gz";
+        let crate_file_path_full = tempdir_handle
+            .as_ref()
+            .join(&crate_file_prefix)
+            .with_extension(crate_file_extension);
+        let crate_url =
+            format!("https://static.crates.io/crates/{pkgname}/{pkgname}-{pkgver}.crate");
+        let success = Command::new("curl")
+            .arg("--output")
+            .arg(crate_file_path_full)
+            .arg(&crate_url)
+            .status()?
+            .success();
+        match success {
+            true => Ok(CrateFile {
+                tempdir_handle,
+                crate_file_prefix,
+                crate_file_extension,
+                config,
+            }),
+            false => Err(Error::DownloadingCrate { crate_url }),
+        }
+    }
+    /// Get the sha256sum for the downloaded .crate file.
+    // NOTE: Possibly future refactor target is crate::sha256sum
+    pub fn get_sha256sum(&self) -> Result<String, Error> {
+        let crate_file_path = self
+            .tempdir_handle
+            .as_ref()
+            .join(&self.crate_file_prefix)
+            .with_extension(self.crate_file_extension);
+        let bytes = std::fs::read(crate_file_path)?;
```

Comment:
How big do we expect crates to ever be? Would it be beneficial to wrap in a `BufReader`?

Reply:
I believe there is a max crate size of 10mb, but it could be increased in the future. Good suggestion and I will try to action.

Reply:
From my perspective this is blocked as `hmac_256::Hash` functions don't take a BufRead, only an array of bytes.

Reply:
Ah so you need to pull it all in into a single thing anyway. No point in buffered reading then. Fair enough.